### ISSUE TITLE
[BUG] Groupby aggregations are slow

### ISSUE THREAD

**Issue by mrocklin** [rapidsai/cudf#533]
**Describe the bug**
Groupby aggregations appear to be significantly slower than Pandas

**Steps/Code to reproduce bug**

```python
import pandas as pd, cudf
pdf = pd.DataFrame({'x': np.random.random(10000000), 'y': np.random.randint(0, 10, size=10000000)})
gdf = cudf.DataFrame.from_pandas(pdf)
%time pdf.groupby('y').mean() # around 300ms
%time gdf.groupby('y').mean() # around 3s
```

**Comment by jrhemstad** [comment]
I just tested on the C++ side and a `mean` aggregation on 10 million random 32-bit keys and 32 bit values only takes 140ms.

**Comment by thomcom** [comment]
GDF is faster than PDF for groupby where the number of groups is _larger_ than 100, but has 10x the runtime of PDF for 10 groups, 30 times the runtime of PDF for 5 groups, and 150 times the runtime of PDF for 2 groups, approximately.

This is an algorithmic bug in the CUDA implementation I'm guessing? It is strongly related to #543, but I don't think #543 is the sole cause.

I analyzed the slowness a bit for the sake of triage:

https://www.dropbox.com/s/uagg8186gftjv7c/Screenshot%202019-01-03%2012.39.30.png?raw=1

And here's the data:
```
Rows 10000000
Groups 100000
PDF 584 ms ± 2.03 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
GDF 43.2 ms ± 615 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Rows 10000000
Groups 1000
PDF 276 ms ± 524 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
GDF 52.6 ms ± 824 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Rows 10000000
Groups 100
PDF 228 ms ± 6.76 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
GDF 170 ms ± 485 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Rows 10000000
Groups 10
PDF 216 ms ± 7.56 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
GDF 2.25 s ± 516 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Rows 10000000
Groups 5
PDF 210 ms ± 3.63 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
GDF 6.71 s ± 232 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Rows 10000000
Groups 2
PDF 233 ms ± 23.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
GDF 34.2 s ± 27.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Rows 10000
Groups 100000
PDF 2.14 ms ± 7.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
GDF 11.1 ms ± 94.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Rows 10000
Groups 1000
PDF 1.34 ms ± 9.98 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
GDF 13.5 ms ± 106 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Rows 10000
Groups 100
PDF 1.22 ms ± 6.07 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
GDF 9.02 ms ± 114 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Rows 10000
Groups 10
PDF 1.21 ms ± 5.73 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
GDF 9.56 ms ± 131 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Rows 10000
Groups 5
PDF 1.21 ms ± 5.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
GDF 10.1 ms ± 153 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Rows 10000
Groups 2
PDF 1.12 ms ± 5.73 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
GDF 12.5 ms ± 115 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by jrhemstad** [comment]
Now that https://github.com/rapidsai/cudf/pull/753 is merged, could @thomcom  or @mrocklin verify that this is no longer an issue?

**Comment by thomcom** [comment]
Yup! Verified that this is fixed in `branch-0.6`. cudf is still slower than pandas for the small dataframe groupbys, because of the setup and teardown cost of kernels etc I expect.

In the previously most difficult case, when cudf took 34s to group 10m rows into 2 groups and Pandas took 233ms, cudf under branch-0.6 now takes 41ms. 👍
